### PR TITLE
updates in genStudies

### PR DESCRIPTION
**Pull request by tyjyang**

### CODE REVIEW

File: scripts/histmakers/mw_with_mu_eta_pt.py [hunk 1]
```diff
@@ -29,8 +29,7 @@
 muon_prefiring_helper, muon_prefiring_helper_stat, muon_prefiring_helper_syst = wremnants.make_muon_prefiring_helpers(era = era)
 scetlibCorrZ_helper = wremnants.makeScetlibCorrHelper(isW=False)
 scetlibCorrW_helper = wremnants.makeScetlibCorrHelper(isW=True)
-qcdScaleByHelicity_Zhelper = wremnants.makeQCDScaleByHelicityHelper(is_w_like = True)
-qcdScaleByHelicity_Whelper = wremnants.makeQCDScaleByHelicityHelper()
+qcdScaleByHelicity_helper = wremnants.makeQCDScaleByHelicityHelper()
```
Comment (kdlong) [inline scripts/histmakers/mw_with_mu_eta_pt.py:32]: I think you want to revert this change

File: scripts/histmakers/mw_with_mu_eta_pt.py [hunk 2]
```diff
@@ -140,16 +138,14 @@ def build_graph(df, dataset):
...
         weight_expr = "weight*weight_pu*weight_fullMuonSF_withTrackingReco*weight_newMuonPrefiringSF"
         if isW or isZ:
             df = wremnants.define_prefsr_vars(df)
             if applyScetlibCorr:
                 df = theory_tools.define_scetlib_corr(df, weight_expr, scetlibCorrZ_helper if isZ else scetlibCorrW_helper)
                 results.extend(theory_tools.make_scetlibCorr_hists(df, "nominal", axes=nominal_axes, cols=nominal_cols, 
-                    helper=scetlibCorrZ_helper if isZ else scetlibCorrW_helper))
-            else:
-                df = df.Define("nominal_weight", weight_expr)
+                    scetlibCorr_helper=scetlibCorrZ_helper if isZ else scetlibCorrW_helper))
```
Comment (kdlong) [inline scripts/histmakers/mw_with_mu_eta_pt.py:148]: Also this one

File: scripts/histmakers/w_z_gen_dists.py [hunk 3]
```diff
@@ -38,8 +35,11 @@
     [0, 0.25, 0.5, 0.75, 1, 1.25, 1.5, 1.75, 2, 2.25, 2.5, 2.75, 3, 3.25, 3.5, 3.75, 4, 10], name = "absYVgen"
 )
 axis_ptVgen = hist.axis.Variable(
-    list(range(41))+[45, 50, 55, 60, 75, 100], name = "ptVgen"
+    [0, 2, 3, 4, 4.75, 5.5, 6.5, 8, 9, 10, 12, 14, 16, 18, 20, 23, 27, 32, 40, 55, 100], name = "ptVgen"
```
Comment (kdlong) [inline scripts/histmakers/w_z_gen_dists.py:38]: I don't care a ton on the binning, but we should just define ptVgen once

**Comment by tyjyang** [comment]
fixed the errors as suggested

File: scripts/histmakers/w_z_gen_dists.py [hunk 4]
```diff
@@ -48,18 +48,51 @@
...
 
 def build_graph(df, dataset):
     print("build graph")
     print(dataset.name)
     results = []
+    
+    if dataset.is_data:
+        df = df.DefinePerSample("weight", "1.0")
+    else:
+        if "h2" in dataset.name:
```
Comment (kdlong) [inline scripts/histmakers/w_z_gen_dists.py:90]: Probably should give this a more descriptive name (with reweighing in it or something)

File: scripts/histmakers/w_z_gen_dists.py [hunk 5]
```diff
@@ -48,18 +48,51 @@
...
+loading_time = time_loading_finished - time_start
+print("time it takes to load datasets: ", loading_time)
 
 
 def build_graph(df, dataset):
     print("build graph")
     print(dataset.name)
     results = []
+    
+    if dataset.is_data:
```
Comment (kdlong) [inline scripts/histmakers/w_z_gen_dists.py:87]: I guess this doesn't hurt anything but not sure we ever want to run the data

File: wremnants/datasets/datasets2016.py [hunk 6]
```diff
@@ -92,8 +143,24 @@ def getDatasets(maxFiles=-1, filt=None):
             xsec = 75.8,
     )
 
-    allPostVFP = [dataPostVFP, WpmunuPostVFP, WmmunuPostVFP, WptaunuPostVFP, WmtaunuPostVFP, ZmmPostVFP, 
-        ZttPostVFP, ttbarlnuPostVFP, ttbarlqPostVFP, wwPostVFP]
+    allPostVFP = [
```
Comment (kdlong) [inline wremnants/datasets/datasets2016.py:146]: I think it would be better to introduce a new list that is only used for the gen, since I think modifying this one could mess up the reco analysis

File: wremnants/histselections.py [hunk 7]
```diff
@@ -24,6 +24,15 @@ def fakeHistIsoRegionIntGen(h, scale=1.):
     print("Slicing")
     return h[{"iso" : 0, "qTgen" : s[::hist.sum]}]
 
+def fakeHistIsoRegionIntGen(h, scale=1.):
```
Comment (kdlong) [inline wremnants/histselections.py:27]: Is this something I did for the low pileup or did you implement this? If it was me, I don't quite remember the use case, so probably better to remove it and restore it when/if it's needed.

File: wremnants/plot_tools.py [hunk 8]
```diff
@@ -7,37 +7,36 @@
 import numpy as np
 hep.style.use(hep.style.ROOT)
 
-def figureWithRatio(href, xlabel, ylabel, ylim, rlabel, rrange, xlim=None):
```
Comment (kdlong) [inline wremnants/plot_tools.py:10]: Is this an intentional change?

File: wremnants/plot_tools.py [hunk 9]
```diff
@@ -7,37 +7,36 @@
...
+def figureWithRatio(href, xlabel, ylabel, ylim, rlabel, rrange):
     hax = href.axes[0]
     width = math.ceil(hax.size/400)
+    print(width)
```
Comment (kdlong) [inline wremnants/plot_tools.py:13]: delete

File: wremnants/plot_tools.py [hunk 10]
```diff
@@ -7,37 +7,36 @@
...
     fig = plt.figure(figsize=(8*width,8))
     ax1 = fig.add_subplot(4, 1, (1, 3)) 
     ax2 = fig.add_subplot(4, 1, 4) 
 
+    ax1.set_xlabel("")
     ax2.set_xlabel(xlabel)
-    ax1.set_xlabel(" ")
```
Comment (kdlong) [inline wremnants/plot_tools.py:18]: I *think* this is needed for the xlabel to actually be removed for the not-ratio axis

File: wremnants/plot_tools.py [hunk 11]
```diff
@@ -50,12 +49,12 @@ def addLegend(ax, ncols=2, extra_text=None):
     ax.legend(handles=handles, labels=labels, prop={'size' : 20*(0.7 if shape == 1 else 1.3)}, ncol=ncols, loc='upper right')
 
 def makeStackPlotWithRatio(histInfo, stackedProcs, label="nominal", unstacked=None, xlabel="", ylabel="Events/bin", 
-                rrange=[0.9, 1.1], ymax=None, xlim=None, binwnorm=None, select={}, nlegcols=2, action=None, extra_text=None):
+                rrange=[0.9, 1.1], ymax=None, xlim=None, binwnorm=None, select={}, action=None, extra_text=None):
```
Comment (kdlong) [inline wremnants/plot_tools.py:52]: I think a few more of these changes are intentional, better to double check it

File: wremnants/plot_tools.py [hunk 12]
```diff
@@ -88,12 +87,18 @@ def makeStackPlotWithRatio(histInfo, stackedProcs, label="nominal", unstacked=No
...
-                rrange=[0.9, 1.1], ymax=None, xlim=None, nlegcols=2, binwnorm=None):
-
+def makePlotWithRatioToRef(hists, labels, colors, xlabel="", ylabel="Events/bin", rlabel="bugfix/bugged",
+                rrange=[0.9, 1.1], ymax=None, xlim=None, binwnorm=None, baseline=None, autorrange=None):
+    if baseline:
+        ratio_plots_starting_idx = 0
+    else:
+        ratio_plots_starting_idx = 1
+    ratio_hists = [hh.divideHists(h, hists[0], cutoff=0.00001) for h in hists[ratio_plots_starting_idx:]]
+#    if autorrange:
```
Comment (kdlong) [inline wremnants/plot_tools.py:100]: Delete commented out

File: wremnants/scetlib_corrections.py [hunk 13]
```diff
@@ -7,29 +7,20 @@
...
-        corrf_Wp = np.load(filename.format(process="Wp"), allow_pickle=True)
-        corrf_Wm = np.load(filename.format(process="Wm"), allow_pickle=True)
-        bins = corrf_Wp["bins"]
-        axis_charge = hist.axis.Regular(2, -2., 2., underflow=False, overflow=False, name = "charge")
-    else:
-        corrf = np.load(filename.format(process="Z"), allow_pickle=True)
-        bins = corrf["bins"]
-        axis_charge = hist.axis.Integer(0, 1, underflow=False, overflow=False, name = "charge")
-
+def makeScetlibCorrHelper(filename=f"{data_dir}/N3LLCorrections/inclusive_{{process}}_pT.npz"):
```
Comment (kdlong) [inline wremnants/scetlib_corrections.py:10]: Unintentional change

File: wremnants/theory_tools.py [hunk 14]
```diff
@@ -65,17 +59,13 @@ def define_and_make_pdf_hists(df, axes, cols, pdfset="nnpdf31"):
...
     return pdfHist, alphaSHist
 
-
 def define_scetlib_corr(df, weight_expr, helper):
     df = df.Define("nominal_weight_uncorr", weight_expr)
-    df = df.Define("scetlibWeight_tensor", helper, ["chargeVgen", "massVgen", "yVgen", "ptVgen", "nominal_weight_uncorr"])
+    df = df.Define("scetlibWeight_tensor", scetlibCorr_helper, ["chargeVgen", "massVgen", "yVgen", "ptVgen", "nominal_weight_uncorr"])
+    scetlibUnc = df.HistoBoost("scetlibUnc", nominal_axes, [*nominal_cols, "scetlibWeight_tensor"], tensor_axes=scetlibCorr_helper.tensor_axes)
     df = df.Define("nominal_weight", "scetlibWeight_tensor(0)")
-    return df
```
Comment (kdlong) [inline wremnants/theory_tools.py:73]: These also look unintentional

File: wremnants/histselections.py (hunk 7 above)
Comment (tyjyang) [reply wremnants/histselections.py:27]: I didn't recall implementing this, will remove for now.

File: wremnants/plot_tools.py (hunk 8 above)
Comment (tyjyang) [reply wremnants/plot_tools.py:10]: This is accidental, reverting the change

File: wremnants/plot_tools.py (hunk 9 above)
Comment (tyjyang) [reply wremnants/plot_tools.py:13]: deleted

File: wremnants/plot_tools.py (hunk 10 above)
Comment (tyjyang) [reply wremnants/plot_tools.py:18]: reverting the change

File: wremnants/plot_tools.py (hunk 12 above)
Comment (tyjyang) [reply wremnants/plot_tools.py:100]: commented line deleted

File: wremnants/scetlib_corrections.py (hunk 13 above)
Comment (tyjyang) [reply wremnants/scetlib_corrections.py:10]: reverted

File: wremnants/theory_tools.py (hunk 14 above)
Comment (tyjyang) [reply wremnants/theory_tools.py:73]: reverted

File: scripts/histmakers/mw_with_mu_eta_pt.py (hunk 1 above)
Comment (tyjyang) [reply scripts/histmakers/mw_with_mu_eta_pt.py]: change reverted

File: scripts/histmakers/w_z_gen_dists.py (hunk 4 above)
Comment (tyjyang) [reply scripts/histmakers/w_z_gen_dists.py:90]: changed sample name to "WplusmunuPostVFP_bugfix_reweight_h2" in both scripts/histmakers/w_z_gen_dists.py and wremnants/datasets/dataset2016.py

File: scripts/histmakers/w_z_gen_dists.py (hunk 5 above)
Comment (tyjyang) [reply scripts/histmakers/w_z_gen_dists.py:87]: Indeed, now data is not loaded by default:
```parser.add_argument("--filterProcs", type=str, nargs="*", help="Only run over processes matched by (subset) of name", default=["Wplus", "Wminus", "Zmumu", "Ztautau"])```
I guess we can keep them here in case we will need it later. Probably need to refactor this a little bit in later iterations so the logic is more clear. But I think it's fine for this PR.

File: wremnants/datasets/datasets2016.py (hunk 6 above)
Comment (tyjyang) [reply wremnants/datasets/datasets2016.py:146]: I made a new list for gen sample comparison. RECO studies would not be affected now

File: wremnants/plot_tools.py (hunk 11 above)
Comment (tyjyang) [reply wremnants/plot_tools.py:51]: indeed, I accidentally deleted the nlegcols argument in 2 functions, where they should be kept. reverted the change.

**Comment by kdlong** [comment]
Maybe a good idea to clear the output of all notebooks before we merge (this can probably be automated but I didn't bother to set it up yet). Otherwise I'm happy to merge.

**Comment by kdlong** [comment]
Maybe a good idea to clear the output of all notebooks before we merge (this can probably be automated but I didn't bother to set it up yet). Otherwise I'm happy to merge.

**Comment by tyjyang** [comment]
good idea! I was able to clear the output automatically by using:
`jupyter nbconvert --clear-output --inplace <notebook.ipynb>`

**Comment by kdlong** [comment]
> good idea! I was able to clear the output automatically by using:
> `jupyter nbconvert --clear-output --inplace <notebook.ipynb>`

But is there a way to make sure this is called at every commit? I think there is, I briefly looked into this, but didn't implement anything

**Comment by bendavid** [comment]
It's possible with "pre-commit hooks"

**Comment by kdlong** [comment]
I'm merging this now because I have some changes on top of it.
